framework/http/FileStream.php

```diff
@@ -0,0 +1,263 @@
+            $resource = fopen(Yii::getAlias($this->filename), $this->mode);
+            if ($resource === false) {
+                throw new InvalidConfigException("Unable to open file '{$this->filename}' with mode '{$this->mode}'");
+            }
+            $this->_resource = $resource;
+        }
+        return $this->_resource;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function __toString()
+    {
+        // __toString cannot throw exception
+        // use trigger_error to bypass this limitation
+        try {
+            $this->seek(0);
+            return $this->getContents();
+        } catch (\Exception $e) {
+            ErrorHandler::convertExceptionToError($e);
+            return '';
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function close()
+    {
+        if ($this->_resource !== null) {
+            fclose($this->_resource);
+            $this->_resource = null;
+            $this->_metadata = null;
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function detach()
+    {
+        if ($this->_resource === null) {
+            return null;
+        }
+        $result = $this->_resource;
+        $this->_resource = null;
+        $this->_metadata = null;
+        return $result;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getSize()
+    {
+        $uri = $this->getMetadata('uri');
+        if (!empty($uri)) {
+            // clear the stat cache in case stream has a URI
+            clearstatcache(true, $uri);
+        }
+
+        $stats = fstat($this->getResource());
+        if (isset($stats['size'])) {
+            return $stats['size'];
+        }
+        return null;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function tell()
+    {
+        $result = ftell($this->getResource());
+        if ($result === false) {
+            throw new \RuntimeException('Unable to determine stream position');
+        }
+        return $result;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function eof()
+    {
+        return feof($this->getResource());
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function isSeekable()
+    {
+        return (bool)$this->getMetadata('seekable');
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function seek($offset, $whence = SEEK_SET)
+    {
+        if (fseek($this->getResource(), $offset, $whence) === -1) {
+            throw new \RuntimeException("Unable to seek to stream position '{$offset}' with whence '{$whence}'");
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function rewind()
+    {
+        $this->seek(0);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function isWritable()
+    {
+        return in_array(
+            $this->getMetadata('mode'),
+            ['w', 'w+', 'rw', 'r+', 'x+', 'c+', 'wb', 'w+b', 'r+b', 'x+b', 'c+b', 'w+t', 'r+t', 'x+t', 'c+t', 'a', 'a+'],
```

Comment:
How about `c`? There are also extra flags such as `e`. 

Either modes should be limited and checked for validity in constructor or checks should include all possible configurations. It worth searching for sub-strings then: `w`, `r+`, `x+`, `c`, `a`.